packages/pg/lib/client.js

```diff
@@ -156,23 +304,311 @@ class Client extends EventEmitter {
     })
   }
 
-  connect(callback) {
+  attachErrorListenerOnClientConnection(client) {
+    client.on('error', () => {
+      if (Client.hostServerInfo.has(client.host)) {
+        Client.failedHosts.set(client.host, Client.hostServerInfo.get(client.host))
+        Client.connectionMap.delete(client.host)
+        Client.hostServerInfo.delete(client.host)
+      }
+      Client.controlClient = undefined
+    })
+  }
+
+  async getConnection() {
+    let currConnectionString = this.connectionString
+    var client = new Client(currConnectionString)
+    this.attachErrorListenerOnClientConnection(client)
+    let lookup = util.promisify(dns.lookup)
+    let addresses = [{ address: client.host }]
+    await lookup(client.host, { family: 0, all: true }).then((res) => {
+      addresses = res
+    })
+    client.host = addresses[0].address // If both resolved then - IPv6 else IPv4
+    client.loadBalance = false
+    client.connectionParameters.loadBalance = false
+    client.topologyKeys = ''
+    client.connectionParameters.topologyKeys = ''
+    if (Client.failedHosts.has(client.host)) {
+      let upHostsList = Client.hostServerInfo.keys()
+      let upHost = upHostsList.next().value
+      client.host = upHost
+    }
+    client.connectionParameters.host = client.host
+    await client.nowConnect().catch(async (err) => {
+      client.connection =
+        client.config.connection ||
+        new Connection({
+          stream: client.config.stream,
+          ssl: client.connectionParameters.ssl,
+          keepAlive: client.config.keepAlive || false,
+          keepAliveInitialDelayMillis: client.config.keepAliveInitialDelayMillis || 0,
+          encoding: client.connectionParameters.client_encoding || 'utf8',
+        })
+      client._connecting = false
+      if (addresses.length === 2) {
+        // If both resolved
+        client.host = addresses[1].address // IPv4
+        if (Client.failedHosts.has(client.host)) {
+          let upHostsList = Client.hostServerInfo.keys()
+          let upHost = upHostsList.next().value
+          client.host = upHost
+        }
+        client.connectionParameters.host = client.host
+        await client.nowConnect()
+      }
+    })
+    return client
+  }
+
+  async getServersInfo() {
+    var client = Client.controlClient
+    var result
+    await client
+      .query(YB_SERVERS_QUERY)
+      .then((res) => {
+        result = res
+      })
+      .catch((err) => {
+        this.getConnection().then(async (res) => {
+          Client.controlClient = res
+          await this.getServersInfo()
+        })
+      })
```

Comment:
Should we handle error condition for getConnection(), similar to how you did in connect() at line#592?